Webhook Reliability: Fetch failed events and process them

htdat opened the pull request:
Fixes #3813

#### Changes proposed in this Pull Request

Add a new service to achieve by doing these things: 

- Hook into `woocommerce_payments_account_refreshed` and check the relevant flag, and decide whether to schedule a job `wcpay_webhook_fetch_events`. Relevant convo p1645601664622659/1645086273.563649-slack-C01BPL3ALGP
- `wcpay_webhook_fetch_events` jobs fetches events from the server, then do two things: 
     - Decide whether to schedule another `wcpay_webhook_fetch_events`.
     - Save data and schedule a job for each event.
- `wcpay_webhook_process_event` jobs simply retrieve data and dispatch to `WC_Payments_Webhook_Processing_Service` (see this PR https://github.com/Automattic/woocommerce-payments/pull/3852)

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

### Testing instructions

Requirement: this server PR 1633-gh-Automattic/woocommerce-payments-server
ActionScheduler management pending page: wp-admin/tools.php?page=action-scheduler&status=pending

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

#### Test case: properly schedule `wcpay_webhook_fetch_events`

**TODOs:** 

- [x] Wait for the server PR to implement this. Otherwise, you can apply this diff for testing 2bd57-pb/#diff - change true/false to see the difference. 
- [x] Update this value `CONTINUOUS_FETCH_FLAG_ACCOUNT_DATA` based on the server PR.

Steps: 

- Follow the server PR, and ensure that your site has some failed not-yet-processed webhook events.
- Go to ActionScheduler management page and run `wcpay_refresh_account_cache`, which forces to retrieve new account data. 
- Expect: a new job `wcpay_webhook_fetch_events` is scheduled. 

#### Test case: run `wcpay_webhook_fetch_events`

- Ensure that your site has some not-yet-processed failed webhook events.
- Go to ActionScheduler management page and run `wcpay_webhook_fetch_events`.
- Expect: failed events are retrieved from the server, and new jobs `wcpay_webhook_process_event` are scheduled with the `event_id` parameter. 
- Try adding more than 10 not-yet-processed failed webhook events. 
- Go to ActionScheduler management page and run `wcpay_webhook_fetch_events`.
- Expect: similar above plus that another `wcpay_webhook_fetch_events` is scheduled (This is to test the `has_more` flag).
- Expect: Go to wp-content/uploads/wc-logs (if you use Docker, it's under the `docker/wordpress` folder), find the log file for today, and see the relevant log entries. Tip: use `event_id` for search.

#### Test case: run `wcpay_webhook_process_event`. 

- Go to ActionScheduler management page and run a `wcpay_webhook_process_event` job.
- Expect: Also check in the log file and see the relevant log entries for `event_id` attached to the job you've just run. 

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) : _Add link here / 'QA Testing Not Applicable'_

htdat committed:
Add skeleton

htdat committed:
Init changes

htdat committed:
First approach:

- Check if having failed wehook events on the server during the account data refresh, and schedule the fetch_event jobs.

- process_event: Build up a request, and dispatch to Webhook_Controller.

RadoslavGeorgiev commented on includes/class-wc-payments-webhook-reliability-service.php:
```diff
@@ -0,0 +1,91 @@
+	private function fetch_events() {
+		$payload = $this->payments_api_client->get_failed_webhook_events();
+
+		$remaining_failed_events = (int) $payload['failed_webhook_count'] ?? 0; // TODO - maybe `failed_webhook_count` (int) or `has_more` (bool). See server PR 1633
```
We will not be using `COUNT(*)` in DB queries, see `$count + 1` in 1633-gh-Automattic/woocommerce-payments-server/files#diff-91becda04a3d12e72f51da7e3f60e14114fc78c8c1090c93af5de85c19b77155R91. `has_more` will simply be boolean in order to know whether to schedule another batch or not.

RadoslavGeorgiev commented on includes/class-wc-payments-webhook-reliability-service.php:
```diff
@@ -0,0 +1,91 @@
+
+	public function process_event( array $event_data ) {
+		// TODO with the following pseudo-code
+		// Verify data?
```
I don't think we need to verify the data here, except for integrity maybe. We're already using the JP's connection package to reach the server, I think we can trust the response.

htdat replied:
> I don't think we need to verify the data here, except for integrity maybe.

Yes, this is what I mean here, not for the security reason.

RadoslavGeorgiev commented on includes/class-wc-payments-webhook-reliability-service.php:
```diff
@@ -0,0 +1,91 @@
+	public function process_event( array $event_data ) {
+		// TODO with the following pseudo-code
+		// Verify data?
+		// Create a request based on this data. Dispatch to Webhook_Controller
```
IMO it would be better to isolate the core event handling from `WC_REST_Payments_Webhook_Controller::handle_webhook()` into a separate method, which can be called here. The code between and including the `woocommerce_payments_before_webhook_delivery` and `wocommerce_payments_after_webhook_delivery` actions seems like a good candidate. The rest (reading REST properties and error handling) should remain there and be specific to standard webhook delivery. This method (`process_event`) should have it's own event handling, as we're not responding to a call from the server here.

htdat replied:
> This method (process_event) should have it's own event handling, as we're not responding to a call from the server here.

Agreed! However, it's tricker in the implementation. I even tried to build an instance of `WC_REST_Payments_Webhook_Controller` then call `handle_webhook()`. However, it's not working as the whole REST API code is not loaded when ActionScheduler runs a job.

> The code between and including the woocommerce_payments_before_webhook_delivery and wocommerce_payments_after_webhook_delivery actions seems like a good candidate.

This is a good point and it may require more refactor to `WC_REST_Payments_Webhook_Controller` so that it can work within an AS job too.

htdat replied:
> That said, it's still possible but it requires more large effort to refactor WC_REST_Payments_Webhook_Controller for our purpose.

Here it is https://github.com/Automattic/woocommerce-payments/pull/3852

htdat committed:
Change names for CONTINUOUS_FETCH_FLAG and use type boolean

htdat committed:
Event storage: change to transient, and use md5 for hashing event_id

htdat committed:
Remove unnecessary cast type.

htdat committed:
Fix PHPUnit errors.

Ref: https://github.com/Automattic/woocommerce-payments/runs/5300698761?check_suite_focus=true

WC_Payments_Payment_Request_Button_Handler_Test::test_get_shipping_options_keeps_chosen_option
TypeError: Argument 1 passed to WC_Payments_Webhook_Reliability_Service::maybe_schedule_fetch_events() must be of the type array, null given, called in /tmp/wordpress/wp-includes/class-wp-hook.php on line 307

htdat committed:
Return false when can not set the transient.

htdat committed:
Remove cast type bool

htdat committed:
Move processing parts of WC_REST_Payments_Webhook_Controller to WC_Payments_Webhook_Processing_Service

htdat committed:
Update loading the new service

htdat committed:
Fix calling wrong function.

htdat committed:
Remove private functions that have been moved into the new service

htdat committed:
Add the missing class

htdat committed:
Add new test class WC_Payments_Webhook_Processing_Service_Test

htdat committed:
Move relevant tests to the new test class

htdat committed:
Fix test: update request_body to event_body, remove $request->set_body

htdat committed:
Fix tests: remove response check for successful cases, check excepctions for failures.

htdat committed:
Update REST terms to general webhook terms

- Rest_Request_Exception > Invalid_Webhook_Data_Exception
- read_rest_property > read_webhook_property

htdat committed:
Load file class-invalid-webhook-data-exception.php

htdat committed:
Extract more code to webhook_processing_service->process

htdat committed:
Fix tests: Move remaining unit tests from Controller test class to Service test class

htdat committed:
Fix test: test_remote_note_fails_returns_response

htdat committed:
Add tests back WC_REST_Payments_Webhook_Controller_Test

htdat committed:
Merging from develop with multiple fixes

htdat committed:
Fix after the merge

See PR 3675

htdat committed:
Fix after the merge for WC_REST_Payments_Webhook_Controller

See PR 3675

htdat committed:
Fix after the merge for WC_Payments_Webhook_Processing_Service_Test

See PR 3675

htdat committed:
Merge branch 'develop' into refactor/decouple-webhook-processing-service

htdat committed:
Miscellaneous changes:

- Update throws tag for WC_Payments_Remote_Note_Service::put_note()
- Move up process_wcpay_notification
- Other small changes

htdat committed:
Add changelog file

htdat committed:
Add missing api_client to WC_Payments_Webhook_Processing_Service

htdat committed:
Merge branch 'develop' into refactor/decouple-webhook-processing-service

htdat committed:
Merge branch 'refactor/decouple-webhook-processing-service' into add/3813-fetch-run-webhooks

# Conflicts:
#	includes/admin/class-wc-rest-payments-webhook-controller.php
#	includes/class-wc-payments.php

htdat committed:
Update to use WC_Payments_Webhook_Processing_Service

htdat committed:
Update get_failed_webhook_events

htdat committed:
Add missing service for self::$webhook_reliability_service

htdat committed:
Re-arrange class WC_Payments_Webhook_Reliability_Service

htdat committed:
Add tests for WC_Payments_Webhook_Reliability_Service

htdat committed:
Properly use continue rather than break

htdat committed:
Update tests for  test_fetch_events_save_data_and_schedule_process_jobs

htdat committed:
Merge branch 'develop' into add/3813-fetch-run-webhooks

htdat committed:
Cosmetic changes

htdat committed:
Add the changelog file.

htdat commented:
Updated the PR description, ready for the first round of review.

htdat committed:
Merge branch 'develop' into refactor/decouple-webhook-processing-service

RadoslavGeorgiev reviewed (approved):
Once again, awesome work, @htdat! 👏 

Immaculate code without any need for changes. The code itself looks excellent and the tests seem to properly cover all needed scenarios.

In terms of testing, I implemented the proper change to the server, which got commited in 1633-gh-Automattic/woocommerce-payments-server/commits/5876c7e8ddb569bde4f00ff95dd5aaa31bc11b39. Once I did that, everything worked flawlessly: 🎊 

0. Temporarily decreased the amount of events returned by the server endpoint to 5 to ease testing.
1. Enabled the setting in the server's `wp-admin` to intentionally store all webhooks as failed.
2. Made a few payments with a card, which automatically triggers disputes. This added up to 11 webhooks in my local database.
3. Checked the orders to confirm that they appear as paid and not disputed (as the webhooks did not reach the client).
4. Triggered the account refresh, which automatically scheduled the fetch action.
5. Triggered the fetch action once, and 5 `process` actions appeared.
6. Triggered those pending actions one by one while refreshing all of those orders, and their notes started appearing one by one, as expected.
7. Repeated the last few steps two more times to fetch the remaining 5 + 1 webhooks. At this point all orders had their disputes processed.
8. Verified that there are no more fetches scheduled.
9. Paid for another order with the same card before triggering the same set of actions. Everything went well, and at the end the slate was clean both for the client, and the server.

The PR is approved, but don't :shipit: yet. I've marked it as blocked until the server gets ready. If for some reason the server is not ready by release time, we can also deploy dummy endpoint(s) for the server, in order to be able to ship client PRs, and activate them later.

htdat committed:
Merge branch 'develop' into refactor/decouple-webhook-processing-service

htdat committed:
Remove TODO as it has been implemented in the server PR.

htdat committed:
Merge branch 'refactor/decouple-webhook-processing-service' into add/3813-fetch-run-webhooks

htdat committed:
Remove TODO as it has been implemented in the server PR.

htdat committed:
Merge branch 'develop' into refactor/decouple-webhook-processing-service

htdat committed:
Fix PHPUnit tests due to this merge #3726

htdat committed:
Merge branch 'refactor/decouple-webhook-processing-service' into add/3813-fetch-run-webhooks

htdat committed:
Fix PHPUnit tests due to this merge #3726

htdat committed:
Merge branch 'develop' into add/3813-fetch-run-webhooks

# Conflicts:
#	includes/class-wc-payments.php

htdat committed:
Merge branch 'develop' into add/3813-fetch-run-webhooks

htdat committed:
Merge branch 'develop' into add/3813-fetch-run-webhooks

deepakpathania commented:
Removing the blocked status from this now since the server PR is ready, we'll merge this soon so that GS can test the other flows with this merged - this itself does not have any testing instructions because, without the server changes, we can't test it. cc @htdat

htdat committed:
Merge branch 'develop' into add/3813-fetch-run-webhooks

htdat committed:
Merge branch 'develop' into add/3813-fetch-run-webhooks
